Compatibility reference page (#31)
scalacenter/scala-3-migration-guide: docs/compatibility.md

Sharing the ABI ensures that Scala 2 and Scala 3 class files can be loaded by the same JVM class loader, and that Scala 2 and Scala 3 `sjsir` files can be linked together by the Scala.js linker. Furthermore it relieves us from a lot of surprising behaviors at runtime.   

## Metaprogramming Incompatibility

A macro definition produces bytecode that will be executed at compile time. When you depend on a macro definition, the compiler loads the class file that contains this definition and then calls it to produce the bytecode of your program, that will be executed at runtime.

The Scala 2 macro mechanisms, which were so far available under the `-experimental` flag, are closely tied to the Scala 2 compiler internals. Therefore it is impossible to execute them in the Dotty compiler.

In contrast, the Scala 3 macros are based on the Tasty format which is designed for stability and compiler independence. We will most certainly have compatibility for the future Scala versions. However, the Scala 2 compiler does not fully support Tasty and thus cannot run Scala 3 macros.

> In order to publish a common macro API for both languages you have to provide the two distinct implementations. A technique for having declaration of Scala 2 macros in a single Scala 3 artifact is being experimented. This is exemplified down below and it will be further detailed in the [macros migration guide](macros.md).

Reviewer: ```suggestion
In order to publish a common macro API for both languages you have to provide the two distinct implementations. A technique for having a declaration of Scala 2 macros in a single Scala 3 artifact is being experimented. This is exemplified down below and it will be further detailed in the [macros migration guide](macros.md).
```

A Scala 3 module can depend on a Scala 2 artifact.

![Scala 3 module depending on a Scala 2 artifact](../images/compatibility/3to2.png)

This diagram can be translated into sbt:

```scala
val scala3 = "0.25.0"
val scala2 = "2.13.3"

lazy val foo = project.in.file("foo")
  .settings(scalaVersion := scala3)
  .dependsOn(bar)

lazy val bar = project.in(file("bar"))
  .settings(
    scalaVersion := scala2,
    crossVersion := CrossVersion.constant("2.13")
  )
```

Note: Setting `crossVersion := CrossVersion.constant(scala3)` is a workaround of [this issue](https://github.com/sbt/sbt/issues/5369).

Or, in case `bar` is a published Scala 2.13 library, we can have:

```scala
val scala3 = "0.25.0"

lazy val foo = project.in.file("foo")
  .settings(
    scalaVersion := scala3,
    libraryDependencies += "org.bar" % "bar_2.13" % "1.0.0"
  )
```

#### The Scala Standard Library

The official standard library for Scala 3.0 is the Scala 2.13 library. Not only the source code is unchanged but it is even not compiled and published under 3.0. It could be, but it would be useless because, as we have seen, a Scala 3.0 module can depend on a Scala 2.13 artifact.

> Therefore the Scala 3 dependency to the standard library follows this exact same pattern.

Reviewer: ```suggestion
Therefore the Scala 3 dependency on the standard library follows this exact same pattern.
```

However those decisions were made for good reasons and by taking care that a good workaround is possible.

In general there is a straightforward cross-compiling solution for every incompatibility, so that the migration from Scala 2 to Scala 3 is easy and smooth.
You can find a corpus of incompatibilities associated with their solutions in the [`incompat/`](../incompat/) folder.

There is an exception though, which is the new metaprogramming framework that replaces the Scala 2 experimental macros.
Further explanations are given down below.

Setting metaprogramming aside, a Scala 2.13 source code can rather easily be translated into a Scala 3 source code.
Once you have it, you will be able to use the new powerful features of Scala 3, which have no equivalent in Scala 2.
Obviously you will lose the source code backward compatibility by doing so.
But amazingly you will still have compatibility at compile time and runtime, which is a breakthrough in the Scala programming history.

## Compile Time Compatibility

In your code you can use public terms and call public methods that are defined in a different module or library.

> It works well as long as the type checker, which is the compiler phase that validates the semantic consistency of the code, is able to read the signatures of those terms and methods, from their class files.

Reviewer 2: ```suggestion
It works well as long as the type checker, which is the compiler phase that validates the semantic consistency of the code, is able to read the signatures of those terms and methods, from the class files containing them.
```

> In Scala 2 they are stored in a dedicated format called the Pickle format.

Reviewer 2: ```suggestion
In Scala 2 the signatures are stored in a dedicated format called the Pickle format.
```

In Scala 3 the story is a bit different because it relies on the TASTy format which is a lot more than a signature layout.
But, for the purpose of migrating from Scala 2 to Scala 3, only the signatures are useful.

### The Dotty Unpicklers

> The first good news is that the Dotty compiler is already able to read both format, the Scala 2 Pickle format and the TASTy format, and thus it can type check code that depends on Scala 2.13 or Scala 3 modules or libraries.

Reviewer 2: ```suggestion
The first good news is that the Dotty compiler is already able to read both formats, the Scala 2 Pickle format and the TASTy format, and thus it can type check code that depends on modules or libraries compiled by Scala versions 2.13 or 3.
```

# Compatibility reference

Scala 3 is a game changer in terms of compatibility in the Scala ecosystem that will greatly improve the day-to-day experience of the Scala programmers, starting from the migration.

In this page you will learn about the compatibility between Scala 2.13 and Scala 3.0 at different levels of the program lifetime.
You will then find how these levels of compatibility can be exploited in concrete examples of dependencies between Scala 2 and Scala 3.

## Compatibility at the Language Level

A large subset of the Scala 2.13 language is still compatible in Scala 3.

> Not all of it though, some constructs have been simplified, restricted or quite simply dropped.

Reviewer 3: ```suggestion
Not all of it though, some constructs have been simplified, restricted or dropped altogether.
```
I don't think "quite simply" is used in English.

Reviewer 2: It is often used to finish off lists of 3, although it's not very formal

However those decisions were made for good reasons and by taking care that a good workaround is possible.

In general there is a straightforward cross-compiling solution for every incompatibility, so that the migration from Scala 2 to Scala 3 is easy and smooth.
You can find a corpus of incompatibilities associated with their solutions in the [`incompat/`](../incompat/) folder.

There is an exception though, which is the new metaprogramming framework that replaces the Scala 2 experimental macros.
Further explanations are given down below.

Setting metaprogramming aside, a Scala 2.13 source code can rather easily be translated into a Scala 3 source code.
Once you have it, you will be able to use the new powerful features of Scala 3, which have no equivalent in Scala 2.

> Obviously you will lose the source code backward compatibility by doing so.

Reviewer 3: ```suggestion
Obviously you will lose the source code forward compatibility by doing so.
```
If you write something with a new version, and it doesn't work with an old version, that's *forward* incompatible. Backward incompatibility is when you write something with an old version and it doesn't work with the newer version.

Author: Thank you for the clarification

> But amazingly you will still have compatibility at compile time and runtime, which is a breakthrough in the Scala programming history.

Reviewer 3: ```suggestion
But amazingly, you will still have compatibility at compile time and runtime, which is a breakthrough in the Scala programming history (except for some of the most exotic features).
```

## Compile Time Compatibility

> In your code you can use public terms and call public methods that are defined in a different module or library.
> It works well as long as the type checker, which is the compiler phase that validates the semantic consistency of the code, is able to read the signatures of those terms and methods, from the class file containing them.

Reviewer 3: ```suggestion
In your code you can use public types and call public methods that are defined in a different module or library.
It works well as long as the type checker, which is the compiler phase that validates the semantic consistency of the code, is able to read the signatures of those types and methods, from the class file containing them.
```
?

Author: Yes of course. It applies for types, terms and methods.

In Scala 2 the signatures are stored in a dedicated format called the Pickle format.
In Scala 3 the story is a bit different because it relies on the TASTy format which is a lot more than a signature layout.
But, for the purpose of migrating from Scala 2 to Scala 3, only the signatures are useful.

### The Dotty Unpicklers

The first good news is that the Dotty compiler is already able to read both formats, the Scala 2 Pickle format and the TASTy format, and thus it can type check code that depends on modules or libraries compiled by Scala 2.13 or Scala 3.

### The Scala 2 TASTy Reader

> The second good news is that the Scala 2 TASTy reader, which enables consuming Scala 3 libraries in Scala 2.13, is well on track.

Reviewer 3: ```suggestion
The second piece of good news is that the Scala 2 TASTy reader, which enables consuming Scala 3 libraries in Scala 2.13, is well on track.
```
"news" is uncountable in English.

It will most certainly be available in Scala 2.13.4 and it will support the following new language features:
- Enums
- Intersection types
- Higher-kinded type lambdas
- Opaque type aliases
- Scala 3 extension methods
- New syntax for context abstraction
- Inheritance of `open` classes and `super` traits
- Exported definitions

You can have a Scala 2 module that depends on a Scala 3 module that uses the new features, and this Scala 3 module can even depend on another Scala 2 module.
Cross-compatibility will not restrain you from using the exciting new features of Scala 3.

In short, we have backward and forward compatibility and so migration can happen gradually and in any order.

## Runtime Compatibility: The Shared ABI

ABI, which stands for Application Binary Interface, is the representation of Scala code in bytecode or Scala.js IR.
It largely determines the runtime behavior of a piece of code.
Scala 2 and Scala 3 share the same ABI.

> A piece of code, provided that the inferred types and implicit resolutions are the same, will produce the same bytecode and ineluctably have the same behavior at runtime.

Reviewer 3: ```suggestion
A piece of code, provided that the inferred types and implicit resolutions are the same, will produce the same bytecode and eventually have the same behavior at runtime.
```
